sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2076,6 +2077,38 @@ class AstBuilder extends DataTypeAstBuilder
       query)
   }
 
+  /**
+   * Add an [[EventTimeWatermark]] to a logical plan.
+   */
+  private def withWatermark(
+      ctx: WatermarkClauseContext,
+      query: LogicalPlan): LogicalPlan = withOrigin(ctx) {
+    val expression = visitNamedExpression(ctx.namedExpression())
+
+    if (expression.isInstanceOf[MultiAlias]) {
+      throw new AnalysisException(
+        errorClass = "CANNOT_USE_MULTI_ALIASES_IN_WATERMARK_CLAUSE",
+        messageParameters = Map()
+      )
+    }
+
+    val namedExpression = expression match {
+      case e: NamedExpression => e
+      case e: Expression => UnresolvedAlias(e)
+    }
+
+    // TODO: This seems to need to find the new expression by index. We can't rely on exprId
+    //  due to UnresolvedAlias. Are there better ways to do?
```

Comment:
I think we need a new logical plan like `UnresolvedEventTimeWatermark`, which will be rewritten into `EventTimeWatermark` + `Project` during analysis.